Various EEBUS improvements

DerAndereAndi opened the pull request:
- Removed EEBUS template and type parameter `forcePVLimits` as it is not needed
- Updated EEBUS reference to https://github.com/evcc-io/eebus/pull/2/commits
- Moved mDNS Library to libp2p/zeroconf (currently a fork with a fix for avahi 0.8 support)
- Accept SKIs with upper letter and spaces
- Show any mDNS entry that claims to support SHIP
- Added support for coordinated charging in ISO EVs. Probably needs adjustments as this in general is a conflict between who controls the charging, but at least this allows evcc to react.
- Improved connection handling with improved automatic reconnects

DerAndereAndi committed:
Add coordinated charging support

DerAndereAndi committed:
Fix linter errors

DerAndereAndi committed:
Minor updates

DerAndereAndi committed:
Improve SKI handling

Also consider spaces and uppercase

DerAndereAndi committed:
Move to libp2p/zeroconf

Using a fork of that until the required avahi patch is merged

DerAndereAndi committed:
Don’t limit mDNS clients to type=EVSE

The assumption that the mDNS TXT type entry conforms to the valid EEBUS device types was wrong

DerAndereAndi committed:
Remove forcePVLimits option

This is not needed

DerAndereAndi committed:
Update module references

Uses eebus fork PR reference

DerAndereAndi committed:
Added Elli charger template

DerAndereAndi committed:
Don’ send limits if communication type is unknown

Only send currents smaller 6A if the communication standard is known otherwise this could cause ISO15118 capable OBCs to stick with IEC61851 when plugging the charge cable in. Or even worse show an error and the cable needs the unplugged, wait for the car to go into sleep and plug it back in.
So if are currentls smaller 6A with unknown communication standard change them to 6A keep in mind, that still will confuse evcc as it thinks charging is stopped, but it isn't yet

DerAndereAndi committed:
Don’t use hardcoded values, use min instead

DerAndereAndi committed:
Update EEBUS ref to improve connection handling

The updated EEBUS repo reference provides improvements for handling EVs with ISO capabilities. Instead of requesting different values in parallel, wait for most until the communication standard is clear. This way we shouldn’t get intermediate load limit values which usage could cause the EV to stick with IEC61851 or show charging errors shortly after plugging it in

DerAndereAndi committed:
Some refactoring and adjustments

DerAndereAndi committed:
Reduce logging

DerAndereAndi committed:
Improve connection handling

- If a remote device disconnects try to reconnect right away with the known mDNS entry. If that fails trigger an mDNS browse again
- If mDNS browse times out, restart it if there are any registered devices not connected

DerAndereAndi committed:
Add safety check and use first element

DerAndereAndi committed:
Add zeroconf mod comment

DerAndereAndi committed:
Merge branch 'master' into feature/eebus-improvements

and update eebus feature branch reference

DerAndereAndi committed:
Update zeroconf mod reference

DerAndereAndi committed:
Fix indexes as it is a map and not an array

DerAndereAndi committed:
Fix mDNS browsing with go1.18.5

DerAndereAndi committed:
Run zeroconf browse continuously

This is required if a connection goes offline and the device comes back at a later time and does not initiate the connection by itself

DerAndereAndi committed:
Update eebus repo reference

DerAndereAndi committed:
Merge branch 'master' into feature/eebus-improvements

# Conflicts:
#	go.mod

DerAndereAndi commented:
Update: This current state is working on my end if the certificate check in go is changed. We are still in progress to get the certificate error fixed. So this is not ready for merging as of now.

DerAndereAndi commented:
@andig Was wäre das gewünschte Verhalten wenn keine Measurements verfügbar sind z.b. weil kein Messgerät verbaut ist. Also für `Currents` und für `ChargedEnergy`. Man wird nicht prüfen können ob so ein Gerät eingebaut sind, sondern nur "warten" ob entsprechende Werte kommen oder nicht.

andig commented:
Ich fang mal von hinten an:

> Momentan wird die geladene Energiemenge nicht zurück gegeben (Bug) 

Gibts Gesamtzählerstand? Der reicht. Sonst rechnen wir das aus der Leistung zusammen falls es die gibt, sonst aus dem eingestellten Strom.

> in der Connect wird einfach die vorgegebene maximale Stromstärke als Messwert zurückgegeben.

Dann gibts auch analog geladener Energie auch keine Strommessung- einfach weglassen. Oder werden wenigstens die Phasen richtig ausgegeben? Das hätte dann schon einen Wert, weiter verarbeitet als für Phasenerkennung werden die in evcc eh nicht.

> Was wäre das gewünschte Verhalten wenn keine Measurements verfügbar sind z.b. weil kein Messgerät verbaut ist. 

Zwei Optionen- das gleiche Problem haben wir bei OCPP auch:
- Anwenderkonfiguration (`hasMeter`: true)
- Erkennung beim Start (z.B. 60s warten) und `decorators` benutzen (da kann ich helfen)

Vielleicht auch Beides- das `hasMeter` könnte dann einfach die Zeit der automatischen Erkennung abkürzen. In jedem Fall muss die Verfügbarkeit beim Start von evcc bekannt sein, das kann leider nicht dynamisch passieren.

Wenn es mal keinen Meßwert geben sollte dann `api.ErrNotAvailable` zurück geben.

andig commented:
Prinzipielle Frage: der Heartbeatfehler ist ja genereller Natur- wie wäre es, den schonmal in einen eigenen PR raus zu ziehen, ebenso die mDNS Themen? Dann bleibts hier übersichtlicher was wirklich elli-spezifisch ist?

andig commented on go.sum:
```diff
@@ -63,6 +63,8 @@ github.com/BurntSushi/toml v1.1.0 h1:ksErzDEI1khOiGPgpwuI7x2ebx/uXQNw7xJpn9Eq1+I
 github.com/BurntSushi/toml v1.1.0/go.mod h1:CxXYINrC8qIiEnFrOxCa7Jy5BFHlXnUU2pbicEuybxQ=
 github.com/BurntSushi/xgb v0.0.0-20160522181843-27f122750802/go.mod h1:IVnqGOEym/WlBOVXweHU+Q+/VP0lqqI8lqeDx9IjBqo=
 github.com/DataDog/datadog-go v3.2.0+incompatible/go.mod h1:LButxg5PwREeZtORoXG3tL4fMGNddJ+vMq1mwgfaqoQ=
+github.com/DerAndereAndi/eebus v0.0.0-20220812150617-4f7442d9e071 h1:W0EymUrYE801OdntqBGkNGIMyi9AS1tsznnehCI07z4=
```
Das ist schräg. Hier müsste

    github.com/evcc-io/eebus 

stehen da es ja ein PR gegen dieses Repo ist???

DerAndereAndi replied:
Es ist ein PR und kein Branch in dem Repo. Also geht das erst wenn der PR dort gemerged ist.

DerAndereAndi commented:
> Gibts Gesamtzählerstand? Der reicht. Sonst rechnen wir das aus der Leistung zusammen falls es die gibt, sonst aus dem eingestellten Strom.

Es gibt keine Zählerstände.

> Dann gibts auch analog geladener Energie auch keine Strommessung- einfach weglassen. Oder werden wenigstens die Phasen richtig ausgegeben? Das hätte dann schon einen Wert, weiter verarbeitet als für Phasenerkennung werden die in evcc eh nicht.

Wenn kein Messgerät vorhanden ist, wird nichts korrekt ausgegeben, sondern immer nur die vorgaben zurückgegeben. Wenn ein Messgerät vorhanden ist, werden die Stromwerte pro Phase korrekt zurückgegeben.
"Einfach weglassen" geht nicht, da man das nicht automatisch erkennen kann ob ein Messgerät nun verbaut ist oder nicht.

> Zwei Optionen- das gleiche Problem haben wir bei OCPP auch:
> 
> * Anwenderkonfiguration (`hasMeter`: true)
> * Erkennung beim Start (z.B. 60s warten) und `decorators` benutzen (da kann ich helfen)
> 
> Vielleicht auch Beides- das `hasMeter` könnte dann einfach die Zeit der automatischen Erkennung abkürzen. In jedem Fall muss die Verfügbarkeit beim Start von evcc bekannt sein, das kann leider nicht dynamisch passieren.
> 
> Wenn es mal keinen Meßwert geben sollte dann `api.ErrNotAvailable` zurück geben.

Ich schau mir das mal an.

DerAndereAndi commented:
> Prinzipielle Frage: der Heartbeatfehler ist ja genereller Natur- wie wäre es, den schonmal in einen eigenen PR raus zu ziehen, ebenso die mDNS Themen? Dann bleibts hier übersichtlicher was wirklich elli-spezifisch ist?

Im Porsche Charger stört der Fehler nicht, also ist das Problem nicht relevant. Es gibt ja bisher keine weiteren EEBUS Wallboxen.

Außerdem sind die Änderungen nicht alle Elli spezifisch, es sind auch Anpassungen für den Porsche Charger mit Hybrid enthalten. Daher ist der PR auch im Titel und der Beschreibung nicht Elli spezifisch.

andig commented:
> Außerdem sind die Änderungen nicht alle Elli spezifisch, es sind auch Anpassungen für den Porsche Charger mit Hybrid enthalten. Daher ist der PR auch im Titel und der Beschreibung nicht Elli spezifisch.

Deshalb könnten wir die auch gerne vorab mergen.

DerAndereAndi commented:
Ich möchte diesen PR nicht komplett auseinandernehmen. Maximal die Templates raus und dann damit und weiteren Anpassungen einen neuen PR.

andig commented:
Templates gerne als eigenen PR. Der `ski` Parameter kann in die `defaults.yaml` um Doppelungen zu vermeiden. Das `forcePVLimits` is auch im backend noch drin- da ebenfalls raus?

DerAndereAndi committed:
Remove elli Wallbox template

Removing from this PR so it can be merged and Elli support will be continued in another PR

DerAndereAndi committed:
Merge branch 'master' into feature/eebus-improvements

DerAndereAndi committed:
Update generated docs

DerAndereAndi commented:
Elli template wurde entfernt, master wurde in den Branch gemerged. Elli support geht dann in einem neuen PR weiter.

DerAndereAndi committed:
Use evcc eebus repo
